pkg/router/controller/route_secret_manager.go

```diff
@@ -118,106 +192,199 @@ func (p *RouteSecretManager) HandleRoute(eventType watch.EventType, route *route
+		return err
+	}
+	return nil
+}
+
+// populateRouteTLSFromSecret updates the TLS configuration of the route using data from the referenced secret.
+// If fetching the secret fails, it records the route rejection and triggers
+// the deletion of the route by calling the HandleRoute method with a watch.Deleted event.
+// Note: This function performs an in-place update of the route. The caller should be aware that the route's TLS configuration will be modified directly.
+func (p *RouteSecretManager) populateRouteTLSFromSecret(route *routev1.Route) error {
+	// read referenced secret
+	secret, err := p.secretManager.GetSecret(context.TODO(), route.Namespace, route.Name)
+	if err != nil {
+		log.Error(err, "failed to get referenced secret")
+		p.recorder.RecordRouteRejection(route, "ExternalCertificateGetFailed", err.Error())
+		p.plugin.HandleRoute(watch.Deleted, route)
+		return err
+	}
+
+	// Update the tls.Certificate and tls.Key fields of the route with the data from the referenced secret.
+	// Since externalCertificate does not contain the CACertificate, tls.CACertificate will not be updated.
+	// NOTE that this update is only performed in-memory and will not reflect in the actual route resource stored in etcd, because
+	// the router does not make kube-client calls to directly update route resources.
+	route.Spec.TLS.Certificate = string(secret.Data["tls.crt"])
+	route.Spec.TLS.Key = string(secret.Data["tls.key"])
+
+	return nil
+}
+
+// unregister removes the route's registration with the secret manager and ensures
+// that any references to the deletedSecrets are cleaned up.
+func (p *RouteSecretManager) unregister(route *routev1.Route) error {
+	// unregister associated secret monitor
+	if err := p.secretManager.UnregisterRoute(route.Namespace, route.Name); err != nil {
+		log.Error(err, "failed to unregister route")
+		return err
+	}
+	// clean the route if present inside deletedSecrets
+	// this is required for the scenario when the associated secret is deleted, before unregistering with secretManager
+	p.deletedSecrets.Delete(generateKey(route.Namespace, route.Name))
+	return nil
+}
+
+// hasExternalCertificate checks whether the given route has an externalCertificate specified.
 func hasExternalCertificate(route *routev1.Route) bool {
 	tls := route.Spec.TLS
 	return tls != nil && tls.ExternalCertificate != nil && len(tls.ExternalCertificate.Name) > 0
 }
+
+// generateKey creates a unique identifier for a route
+func generateKey(namespace, routeName string) string {
+	return fmt.Sprintf("%s/%s", namespace, routeName)
+}
+
+// getUpdatedRoute fetches the latest version of the route using the RouteLister.
+func (p *RouteSecretManager) getUpdatedRoute(namespace, routeName string) (*routev1.Route, error) {
+	updatedRoute, err := p.routelister.Routes(namespace).Get(routeName)
+	if err != nil {
+		log.Error(err, "failed to get route", "namespace", namespace, "route", routeName)
+		return nil, err
+	}
+	return updatedRoute, nil
```

Comment:
If you follow the advice to log the error in the places of their usage, this function can be simplified to:
```suggestion
	return p.routelister.Routes(namespace).Get(routeName)
```
As a matter of fact this function would not be needed at all `p.routelister.Routes(namespace).Get(routeName)` can be used directly where it's needed.

Reply:
Thanks for the suggestion. I've removed this function and using `p.routelister.Routes(namespace).Get(routeName)` directly.